File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    const result = [];
    
    if (p.length > s.length) return result;
    
    const pCount = new Array(26).fill(0);
    const windowCount = new Array(26).fill(0);
    
    for (let i = 0; i < p.length; i++) {
        pCount[p.charCodeAt(i) - 97]++;
        windowCount[s.charCodeAt(i) - 97]++;
    }
    
    let matches = 0;
    for (let i = 0; i < 26; i++) {
        if (pCount[i] === windowCount[i]) {
            matches++;
        }
    }
    
    // Check first window
    if (matches === 26) {
        result.push(0);
    }
    
    // Slide the window
    for (let i = p.length; i < s.length; i++) {
        const rightIndex = s.charCodeAt(i) - 97;
        windowCount[rightIndex]++;
        
        if (windowCount[rightIndex] === pCount[rightIndex]) {
            matches++;
        } else if (windowCount[rightIndex] - 1 === pCount[rightIndex]) {
            matches--;
        }
        
        const leftIndex = s.charCodeAt(i - p.length) - 97;
        windowCount[leftIndex]--;
        
        if (windowCount[leftIndex] === pCount[leftIndex]) {
            matches++;
        } else if (windowCount[leftIndex] + 1 === pCount[leftIndex]) {
            matches--;
        }
        
        if (matches === 26) {
            result.push(i - p.length + 1);
        }
    }
    
    return result;
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        result = []
        len_s, len_p = len(s), len(p)
        
        if len_s < len_p:
            return result
        
        p_count = Counter(p)
        window_count = Counter(s[:len_p])
        
        if p_count == window_count:
            result.append(0)
        
        for i in range(len_p, len_s):
            window_count[s[i]] += 1
            
            left_char = s[i - len_p]
            window_count[left_char] -= 1
            if window_count[left_char] == 0:
                del window_count[left_char]
            
            if p_count == window_count:
                result.append(i - len_p + 1)
        
        return result

